kubetest2-gke/deployer/deployer.go

```diff
@@ -218,61 +227,156 @@ func (d *deployer) Build() error {
+		networkHostProject := d.projects[0]
+
+		// Disassociate the rest of the projects
+		for i := 1; i < len(projects); i++ {
+			if err := runWithOutput(exec.Command("gcloud", "compute", "shared-vpc",
+				"associated-projects", "remove", projects[i],
+				"--host-project", networkHostProject)); err != nil {
+				return fmt.Errorf("error associating project (%s) to Shared VPC: %s", projects[i], err)
+			}
+		}
+
+		// Disable Shared VPC for multiproject requests on the host project
+		if err := runWithOutput(exec.Command("gcloud", "compute", "shared-vpc", "disable", networkHostProject)); err != nil {
+			return fmt.Errorf("error creating Shared VPC host project: %s", err)
+		}
 	}
 
+	return nil
+}
+
+func (d *deployer) createNetwork(projects []string) error {
 	// Create network if it doesn't exist.
+	// Verify networks for all projects and configure the mesh if parameter is set
 	if runWithNoOutput(exec.Command("gcloud", "compute", "networks", "describe", d.network,
-		"--project="+d.project,
+		"--project="+d.projects[0],
 		"--format=value(name)")) != nil {
 		// Assume error implies non-existent.
 		log.Printf("Couldn't describe network '%s', assuming it doesn't exist and creating it", d.network)
 		if err := runWithOutput(exec.Command("gcloud", "compute", "networks", "create", d.network,
-			"--project="+d.project,
+			"--project="+d.projects[0],
 			"--subnet-mode=auto")); err != nil {
 			return err
 		}
 	}
+	return nil
+}
 
-	loc, err := d.location()
-	if err != nil {
+func (d *deployer) transformNetworkName() string {
+	if len(d.projects) == 1 {
+		return d.network
+	}
+	// Multiproject should specify the network at cluster creation such as:
+	// projects/HOST_PROJECT_ID/global/networks/SHARED_VPC_NETWORK
+	return fmt.Sprintf("projects/%s/global/networks/%s", d.projects[0], d.network)
+}
+
+// Deployer implementation methods below
+func (d *deployer) Up() error {
+	if err := d.init(); err != nil {
 		return err
 	}
-	args := make([]string, len(d.createCommand()))
-	copy(args, d.createCommand())
-	args = append(args,
-		"--project="+d.project,
-		loc,
-		"--machine-type="+d.machineType,
-		"--image-type="+image,
-		"--num-nodes="+strconv.Itoa(d.nodes),
-		"--network="+d.network,
-	)
-	fmt.Printf("Environment: %v", os.Environ())
-
-	args = append(args, d.cluster)
-	fmt.Printf("Gcloud command: gcloud %+v", args)
-	if err := runWithOutput(exec.Command("gcloud", args...)); err != nil {
-		return fmt.Errorf("error creating cluster: %v", err)
+
+	if err := d.enableSharedVPCProjects(d.projects); err != nil {
+		return err
 	}
+
+	if err := d.createNetwork(d.projects); err != nil {
+		return err
+	}
+
+	for _, project := range d.projects {
+		// TODO Prepare should be performed per GCP project
+		if err := d.prepareGcpIfNeeded(project); err != nil {
+			return err
+		}
+
+		loc, err := d.location()
+		if err != nil {
+			return err
+		}
+
+		for _, cluster := range d.clustersLayout[project] {
```

Comment:
multi-cluster creation should definitely be parallelized!

Reply:
Definitely agree.
Changed to use `errgroup` to parallelize the operations.